src/language-markdown/printer-markdown.js

```diff
@@ -549,48 +551,74 @@ function printTable(path, options, print) {
     contents.push(rowContents);
   }, "children");
 
+  // Get the width of each column
   const columnMaxWidths = contents.reduce(
     (currentWidths, rowContents) =>
       currentWidths.map((width, columnIndex) =>
         Math.max(width, privateUtil.getStringWidth(rowContents[columnIndex]))
       ),
     contents[0].map(() => 3) // minimum width = 3 (---, :--, :-:, --:)
   );
-
-  return join(hardline, [
+  const alignedTable = join(hardlineWithoutBreakParent, [
     printRow(contents[0]),
     printSeparator(),
-    join(hardline, contents.slice(1).map(printRow))
+    join(
+      hardlineWithoutBreakParent,
+      contents.slice(1).map(rowContents => printRow(rowContents))
+    )
   ]);
 
-  function printSeparator() {
+  if (options.proseWrap !== "never") {
+    return alignedTable;
+  }
+
+  // Only if the --prose-wrap never is set and it exceeds the print width.
+  const compactTable = join(hardlineWithoutBreakParent, [
+    // isCompact
+    printRow(contents[0], /* isCompact */ true),
+    printSeparator(/* isCompact */ true),
+    join(
+      hardlineWithoutBreakParent,
+      contents
+        .slice(1)
+        .map(rowContents => printRow(rowContents, /* isCompact */ true))
+    )
+  ]);
+
+  return group(ifBreak(compactTable, alignedTable));
```

Comment:
Since we used the `hardlineWithoutBreakParent` trick, we should add a `breakParent` here as well so that the output is always consistent (to avoid future errors).

```suggestion
  return concat([breakParent, group(ifBreak(compactTable, alignedTable))]);
```